### PR TITLE
JP-3350: Set all DO_NOT_USE pixels to NaN after flat field

### LINKED ISSUE #7834: Set DO_NOT_USE pixels to NaN after flat field
_Issue [JP-3350](https://jira.stsci.edu/browse/JP-3350) was created on JIRA by [Melanie Clarke](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mclarke):_

NIRSpec flat fields incorporate a flux calibration factor from DN/s to MJy/pixel, which means that the flux values change by many orders of magnitude after flat fielding, from around 1e2 to around 1e-9.

During the flat fielding process, some pixels may be left uncorrected because there is no applicable flat data.  These are marked in the DQ plane with NO_FLAT_FIELD and DO_NOT_USE flags, so they are not used in further processing, but they remain in the image.  This can cause some confusion for users looking at intermediate products, before resampling: the scale is so widely different for the corrected and uncorrected pixels, that it can throw off automated scaling algorithms, making it look like there is no spectral trace in the image.  

The workaround for users is to set pixels marked DO_NOT_USE to NaN in the intermediate products.  It would be helpful if the pipeline could do this automatically: at the end of the flat field step, any new DO_NOT_USE pixels should be set to NaN.

### PR DESCRIPTION
**Pull request by melanieclarke**
Resolves [JP-3350](https://jira.stsci.edu/browse/JP-3350)

Closes #7834 

For NIRSpec data, any new DO_NOT_USE pixels generated during the flat field process are set to NaN. This addresses a cosmetic issue, where data that has not been flat-fielded is many orders of magnitude higher than data that has been corrected.  The pixels are already ignored in further processing, but can be confusing to look at in intermediate data products.

**Checklist for maintainers**
- [x] added entry in `CHANGES.rst` within the relevant release section
- [ ] updated or added relevant tests
- [ ] updated relevant documentation
- [x] added relevant milestone
- [x] added relevant label(s)
- [x] ran regression tests, post a link to the Jenkins job below.
      [How to run regression tests on a PR](https://github.com/spacetelescope/jwst/wiki/Running-Regression-Tests-Against-PR-Branches)
- [ ] Make sure the JIRA ticket is [resolved properly](https://github.com/spacetelescope/jwst/wiki/How-to-resolve-JIRA-issues)

### CODE REVIEW

**Comment by hbushouse** [comment]
Regression tests started at https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/981

**Review by hbushouse — COMMENTED** [review]
Looks good, but one question about whether this is really only resetting pixels that are flagged from the flats, as opposed to all DO_NOT_USE pixels in the science data.

File: jwst/flatfield/flat_field.py [hunk 1]
```diff
@@ -440,6 +440,10 @@ def nirspec_fs_msa(output_model, f_flat_model, s_flat_model, d_flat_model, dispa
         # Combine the science and flat DQ arrays
         slit.dq |= slit_flat.dq
 
+        # Make sure any new DO_NOT_USE pixels are NaN
+        dnu = np.bitwise_and(slit.dq, dqflags.pixel['DO_NOT_USE'])
```
[the remark this answers is not in the document]
Comment (melanieclarke) [reply jwst/flatfield/flat_field.py:444]: That's true - this code sets all DO_NOT_USE pixels to NaN.  I think the only ones that are not already set to NaN at this point in the pipeline are the ones newly flagged by this step, but the comment is misleading.  I can update it.

I'll also make the change to combine the two lines of code and make one less variable.

**Comment by hbushouse** [comment]
Regression test results definitely show lots of differences from the flat_field and downstream steps: https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/981/testReport/

**Comment by melanieclarke** [comment]
I would expect changes to all intermediate products after flat field. I would not expect changes to resampled or extracted products, since those should be ignoring DO_NOT_USE pixels anyway.

I see there are some changes to MOS s2d and x1d files, but I think those are from the other recent change, to widen the MOS cutouts.

**Review by hbushouse — APPROVED** [review]

**Pull request merged**